File: src/test/java/no/novari/flyt/history/repository/utils/BatchPersister.java
package no.novari.flyt.history.repository.utils;

import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.history.repository.utils.performance.Timer;
import org.springframework.data.jpa.repository.JpaRepository;
import org.testcontainers.shaded.com.google.common.collect.Lists;

import java.time.Duration;
import java.util.List;
import java.util.stream.IntStream;

import static no.novari.flyt.history.repository.utils.performance.DurationFormatter.formatDuration;

@Slf4j
public class BatchPersister<T> {

    private final JpaRepository<T, Long> repository;
    private final int batchSize;

    public BatchPersister(
            JpaRepository<T, Long> repository,
            int batchSize
    ) {
        this.repository = repository;
        this.batchSize = batchSize;
    }

    public void persistInBatches(List<T> entities) {
        int numberOfEntities = entities.size();
        List<List<T>> entityBatches = Lists.partition(entities, batchSize);
        int numberOfBatches = entityBatches.size();

        log.debug("Persisting {} entities in {} batches of size {}", numberOfEntities, numberOfBatches, batchSize);
        Timer timer = Timer.start();

        IntStream.range(0, numberOfBatches)
                .forEach(i -> {
                    List<T> batchEntities = entityBatches.get(i);
                    Timer batchTimer = Timer.start();
                    repository.saveAllAndFlush(batchEntities);
                    Duration batchElapsedTime = batchTimer.getElapsedTime();
                    log.debug("Persisted batch {} of {} in {} ({}/s)",
                            i + 1,
                            numberOfBatches,
                            formatDuration(batchElapsedTime),
                            ((long) batchEntities.size() * 1000) / batchElapsedTime.toMillis()
                    );
                });
        Duration elapsedTime = timer.getElapsedTime();
        log.debug("Persisted {} entities in {} ({}/s)",
                numberOfEntities,
                formatDuration(elapsedTime),
                String.format("%.2f", (double) numberOfEntities * 1000 / elapsedTime.toMillis())
        );
    }

}

File: src/test/kotlin/no/novari/flyt/history/repository/utils/BatchPersister.kt
package no.novari.flyt.history.repository.utils

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.flyt.history.repository.utils.performance.DurationFormatter.formatDuration
import no.novari.flyt.history.repository.utils.performance.Timer
import org.springframework.data.jpa.repository.JpaRepository

class BatchPersister<T>(
    private val repository: JpaRepository<T, Long>,
    private val batchSize: Int,
) {
    fun persistInBatches(entities: List<T>) {
        val numberOfEntities = entities.size
        val entityBatches = entities.chunked(batchSize)
        val numberOfBatches = entityBatches.size

        log.atDebug {
            message = "Persisting {} entities in {} batches of size {}"
            arguments = arrayOf(numberOfEntities, numberOfBatches, batchSize)
        }
        val timer = Timer.start()

        entityBatches.forEachIndexed { index, batchEntities ->
            val batchTimer = Timer.start()
            repository.saveAllAndFlush(batchEntities)
            val batchElapsedTime = batchTimer.elapsedTime
            log.atDebug {
                message = "Persisted batch {} of {} in {} ({}/s)"
                arguments =
                    arrayOf(
                        index + 1,
                        numberOfBatches,
                        formatDuration(batchElapsedTime),
                        ((batchEntities.size.toLong()) * 1000) / batchElapsedTime.toMillis(),
                    )
            }
        }

        val elapsedTime = timer.elapsedTime
        log.atDebug {
            message = "Persisted {} entities in {} ({}/s)"
            arguments =
                arrayOf(
                    numberOfEntities,
                    formatDuration(elapsedTime),
                    String.format("%.2f", numberOfEntities.toDouble() * 1000 / elapsedTime.toMillis()),
                )
        }
    }

    private companion object {
        private val log = KotlinLogging.logger {}
    }
}
